### PR TITLE
feat: ADA conflict detection & admin resolution

### PR DESCRIPTION
**Pull request by kouloumos**
## Summary

- When polling for Diavgeia decisions, the same ADA can match subjects across different meetings. Previously this caused a P2002 unique constraint error, rolling back the entire transaction and losing all matches from that poll.
- Now, conflicting subjects record a `claimedAda` instead of failing. Non-conflicting matches in the same poll are still saved normally.
- Admin resolves each conflict via "Reassign" (move the decision) or "Dismiss" (discard the claim) in the Diavgeia admin page, with a confirmation dialog.
- Multiple subjects can independently claim the same ADA — the table groups them by ADA with rowSpan.

## Test plan

Run integration tests: `npm run test:integration -- tests/integration/task-handlers.test.ts`

Trigger polls for two meetings with overlapping subjects to verify conflicts are detected while non-conflicting matches are still saved. Visit `/admin/diavgeia` to verify the Conflicts section appears. Test both Reassign and Dismiss resolution flows.

---

> [!NOTE]
> **Medium Risk**
> Touches decision polling and database writes around the `Decision.ada` unique constraint; incorrect conflict handling or resolution could misassign or drop decision links across subjects. Changes are guarded with transactions and integration tests but still affect production data integrity.
> 
> **Overview**
> Prevents `pollDecisions` runs from failing on `Decision.ada` uniqueness collisions by adding a nullable `Subject.claimedAda` and recording conflicts (skip decision upsert, set claim) while continuing to save non-conflicting matches in the same poll; also clears stale claims on successful upserts and logs conflict counts.
> 
> Adds an admin-facing **ADA Conflicts** section to the Diavgeia polling stats page to review grouped conflicts and resolve each via a confirmation dialog (*Reassign* moves the decision to the claiming subject; *Dismiss* clears the claim), backed by a new `resolveAdaConflict` server action with auth checks, transactional move, and cross-city guard.
> 
> Improves polling stats filters by expanding city/meeting dropdown sources (union of task history + still-polling meetings) and applying city/meeting filtering consistently across still-polling meetings, recent polls, and conflicts; adds integration test coverage for conflict detection, claim lifecycle, and resolution paths.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 49ba069915e4d5a79ed9971035159032e2e42c31. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

This PR prevents poll runs from failing entirely when the same ADA is matched to subjects across different meetings. Instead of letting the P2002 unique-constraint violation roll back the whole transaction, conflicts are recorded as `Subject.claimedAda` and an admin resolution UI is added to `/admin/diavgeia`.

**Key changes:**
- **Migration & schema** — minimal nullable `claimedAda` column addition
- **`handlePollDecisionsResult`** — conflict detection uses a pre-built in-memory map (updated after each successful upsert), and a P2002 catch provides a safety net for concurrent polls
- **`resolveAdaConflict`** — reassign is fully transactional with a fresh re-read, defensive cross-city check, and proper `taskId`/`createdById` preservation; dismiss is a single-update write
- **`PollingStats` / `ConflictsSection`** — adds ADA-grouped conflict table with confirmation dialog and error banner; uses ADA-level `resolvingAda` state to prevent concurrent UI actions within an ADA group
- **Tests** — comprehensive integration tests cover conflict detection, both resolution paths, concurrent-poll fallback, and edge cases

All previously-flagged critical issues have been addressed across multiple force-pushes: `taskId` preservation on reassign, error feedback in the UI, stale claim clearing after successful upsert, auth-before-data-read ordering, cross-city reassign guard, and concurrent-poll P2002 race. The code is well-tested and safe to merge.

<h3>Confidence Score: 4/5</h3>

- Safe to merge — core logic is correct, well-tested, and all critical issues have been addressed.
- The PR implements robust conflict detection and resolution with proper transactional safety, auth checks, and guards. The reassign path is fully transactional with re-reads and cross-city validation; the concurrent-poll P2002 race has a fallback catch. Integration tests are comprehensive and cover both resolution flows, edge cases, and the concurrent-poll fallback. No critical logic issues remain.
- No files require special attention

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Poll returns matches] --> B[Build adaToExistingSubject map<br/>from current DB decisions]
    B --> C[For each match]
    C --> D{ADA already<br/>belongs to<br/>different subject?}
    D -- No --> E[decision.upsert]
    E --> F{P2002 concurrent<br/>race?}
    F -- No --> G[Clear stale claimedAda<br/>Update in-memory map]
    F -- Yes --> H[subject.claimedAda = ada<br/>conflictCount++]
    D -- Yes --> H
    H --> C
    G --> C

    subgraph Admin Resolution
        I[Admin visits /admin/diavgeia] --> J{conflicts.length > 0?}
        J -- Yes --> K[Show ADA Conflicts table<br/>grouped by ADA with rowSpan]
        K --> L{Admin chooses}
        L -- Dismiss --> M[subject.claimedAda = null]
        L -- Reassign --> N[Transaction: re-read claimedAda<br/>check existingOnClaiming<br/>cross-city guard<br/>delete old decision<br/>create on claiming subject<br/>clear claimedAda]
        N --> O[Page refreshes via updateParam]
        M --> O
    end
```

<details><summary><h3>Comments Outside Diff (5)</h3></summary>

1. General comment 

   **Intra-poll duplicate ADA still causes P2002 transaction rollback**

   The `adaToExistingSubject` map is built once from the database at the start of the loop (lines 829–833) and is never updated as new matches are processed. If two different subjects in the same poll both match the same brand-new ADA (one that doesn't yet exist in the DB), the second occurrence will pass the conflict check but fail the unique constraint at the upsert, rolling back the entire transaction — the same failure mode this PR aims to fix.

   **Example scenario:**
   1. Subject 1 matches ADA-NEW → no entry in `adaToExistingSubject` → upsert succeeds
   2. Subject 2 matches ADA-NEW → map still undefined (never updated) → conflict check passes → upsert violates unique constraint → P2002 → rollback

   **Suggested fix:** After each successful upsert, update the map so subsequent matches see the newly-claimed ADA:

   ```typescript
   processedCount++;
   // Track this ADA as now belonging to match.subjectId, so
   // later matches in the same poll see it as already taken.
   if (match.ada) {
       adaToExistingSubject.set(match.ada, match.subjectId);
   }
   ```

   Alternatively, pre-scan incoming matches for duplicate ADAs and record `claimedAda` for later occurrences before the loop starts.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/tasks/pollDecisions.ts
   Line: 827-893

   Comment:
   **Intra-poll duplicate ADA still causes P2002 transaction rollback**

   The `adaToExistingSubject` map is built once from the database at the start of the loop (lines 829–833) and is never updated as new matches are processed. If two different subjects in the same poll both match the same brand-new ADA (one that doesn't yet exist in the DB), the second occurrence will pass the conflict check but fail the unique constraint at the upsert, rolling back the entire transaction — the same failure mode this PR aims to fix.

   **Example scenario:**
   1. Subject 1 matches ADA-NEW → no entry in `adaToExistingSubject` → upsert succeeds
   2. Subject 2 matches ADA-NEW → map still undefined (never updated) → conflict check passes → upsert violates unique constraint → P2002 → rollback

   **Suggested fix:** After each successful upsert, update the map so subsequent matches see the newly-claimed ADA:

   ```typescript
   processedCount++;
   // Track this ADA as now belonging to match.subjectId, so
   // later matches in the same poll see it as already taken.
   if (match.ada) {
       adaToExistingSubject.set(match.ada, match.subjectId);
   }
   ```

   Alternatively, pre-scan incoming matches for duplicate ADAs and record `claimedAda` for later occurrences before the loop starts.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. General comment 

   **`discoveries` query ignores `cityId` / `councilMeetingId` filters**

   `getPollingStats` correctly applies the city/meeting filters to the `stillPollingMeetings`, `recentPollTasks`, and `claimingSubjects` queries, but the `discoveries` fetch starting here has no such filter:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: { taskId: { not: null } },
       // ← no cityId / councilMeetingId filter
   ```

   As a result, when an admin selects a city (or meeting) filter in the admin UI, the **"Discoveries"** section and the `totalDiscoveries` summary card always reflect the global data set — inconsistent with every other section on the page.

   The `task` relation carries `cityId`/`councilMeetingId`, so the filter can be pushed into the query:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: {
           taskId: { not: null },
           ...(cityId && { task: { cityId } }),
           ...(councilMeetingId && { task: { councilMeetingId } }),
       },
       ...
   });
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/tasks/pollDecisions.ts
   Line: 195-223

   Comment:
   **`discoveries` query ignores `cityId` / `councilMeetingId` filters**

   `getPollingStats` correctly applies the city/meeting filters to the `stillPollingMeetings`, `recentPollTasks`, and `claimingSubjects` queries, but the `discoveries` fetch starting here has no such filter:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: { taskId: { not: null } },
       // ← no cityId / councilMeetingId filter
   ```

   As a result, when an admin selects a city (or meeting) filter in the admin UI, the **"Discoveries"** section and the `totalDiscoveries` summary card always reflect the global data set — inconsistent with every other section on the page.

   The `task` relation carries `cityId`/`councilMeetingId`, so the filter can be pushed into the query:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: {
           taskId: { not: null },
           ...(cityId && { task: { cityId } }),
           ...(councilMeetingId && { task: { councilMeetingId } }),
       },
       ...
   });
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

3. General comment 

   **"Reassign" silently degrades to "dismiss" when claiming subject already has a decision**

   When `existingOnClaiming` is truthy the inner block is skipped entirely and the function just clears `claimedAda`. From the admin's perspective they clicked **Reassign** (dialog copy: *"This will move the decision from the current subject to the claiming subject"*), but what actually happened is a quiet dismiss — the original decision stays on its current subject and nothing moves.

   There is no exception thrown and no return value, so `handleResolve` in the UI will call `updateParam` to refresh the page. The conflict row disappears and the admin has no indication that the requested reassignment was not performed.

   Consider either:
   - Returning a discriminated result (e.g. `{ outcome: 'skipped' | 'reassigned' }`) and surfacing the skip reason in the UI, or
   - Throwing a descriptive error that the existing `catch` in `handleResolve` can display (`setError`) so the admin knows why the action was a no-op.

   The test is titled `'reassign — when claiming subject already has a decision, **dismisses instead**'` which confirms this is known behaviour — it just needs to be surfaced to the user.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/tasks/pollDecisions.ts
   Line: 688-714

   Comment:
   **"Reassign" silently degrades to "dismiss" when claiming subject already has a decision**

   When `existingOnClaiming` is truthy the inner block is skipped entirely and the function just clears `claimedAda`. From the admin's perspective they clicked **Reassign** (dialog copy: *"This will move the decision from the current subject to the claiming subject"*), but what actually happened is a quiet dismiss — the original decision stays on its current subject and nothing moves.

   There is no exception thrown and no return value, so `handleResolve` in the UI will call `updateParam` to refresh the page. The conflict row disappears and the admin has no indication that the requested reassignment was not performed.

   Consider either:
   - Returning a discriminated result (e.g. `{ outcome: 'skipped' | 'reassigned' }`) and surfacing the skip reason in the UI, or
   - Throwing a descriptive error that the existing `catch` in `handleResolve` can display (`setError`) so the admin knows why the action was a no-op.

   The test is titled `'reassign — when claiming subject already has a decision, **dismisses instead**'` which confirms this is known behaviour — it just needs to be surfaced to the user.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

4. General comment 

   **`claimedAda` is read outside the transaction and used as a closure value inside it**

   `subject.claimedAda` is read before the transaction begins. There is a window (auth check + any I/O latency) in which another concurrent request (e.g. a second admin resolving the same conflict, or a re-poll) could change or clear `claimedAda` on that subject. The transaction then operates on the stale ADA:

   1. `tx.decision.findUnique({ where: { ada: subject.claimedAda! } })` looks up the **old** ADA
   2. Deletes + recreates the decision for that old ADA
   3. `tx.subject.update({ data: { claimedAda: null } })` clears whatever value is now in the DB — possibly a *different* claimed ADA that was set by the concurrent poll

   In an admin panel this race is uncommon, but since the function is already inside a transaction, re-reading `claimedAda` inside the transaction (with appropriate serialisation) would be the correct fix:

   ```typescript
   await prisma.$transaction(async (tx) => {
       const freshSubject = await tx.subject.findUnique({
           where: { id: subjectId },
           select: { claimedAda: true },
       });
       if (!freshSubject?.claimedAda) return; // already resolved concurrently

       const existingOnClaiming = await tx.decision.findUnique({ where: { subjectId } });
       if (!existingOnClaiming) {
           const existingDecision = await tx.decision.findUnique({
               where: { ada: freshSubject.claimedAda },
           });
           // ... rest unchanged
       }
       await tx.subject.update({ where: { id: subjectId }, data: { claimedAda: null } });
   });
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/tasks/pollDecisions.ts
   Line: 667-696

   Comment:
   **`claimedAda` is read outside the transaction and used as a closure value inside it**

   `subject.claimedAda` is read before the transaction begins. There is a window (auth check + any I/O latency) in which another concurrent request (e.g. a second admin resolving the same conflict, or a re-poll) could change or clear `claimedAda` on that subject. The transaction then operates on the stale ADA:

   1. `tx.decision.findUnique({ where: { ada: subject.claimedAda! } })` looks up the **old** ADA
   2. Deletes + recreates the decision for that old ADA
   3. `tx.subject.update({ data: { claimedAda: null } })` clears whatever value is now in the DB — possibly a *different* claimed ADA that was set by the concurrent poll

   In an admin panel this race is uncommon, but since the function is already inside a transaction, re-reading `claimedAda` inside the transaction (with appropriate serialisation) would be the correct fix:

   ```typescript
   await prisma.$transaction(async (tx) => {
       const freshSubject = await tx.subject.findUnique({
           where: { id: subjectId },
           select: { claimedAda: true },
       });
       if (!freshSubject?.claimedAda) return; // already resolved concurrently

       const existingOnClaiming = await tx.decision.findUnique({ where: { subjectId } });
       if (!existingOnClaiming) {
           const existingDecision = await tx.decision.findUnique({
               where: { ada: freshSubject.claimedAda },
           });
           // ... rest unchanged
       }
       await tx.subject.update({ where: { id: subjectId }, data: { claimedAda: null } });
   });
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

5. `src/lib/tasks/pollDecisions.ts`, line 197-222 ([link](https://github.com/schemalabz/opencouncil/blob/b6381324c6ba6962c5a14df4ac988402da05bfd7/src/lib/tasks/pollDecisions.ts#L197-L222)) 

   **`discoveries` ignores city/meeting filter parameters**

   `getPollingStats` accepts `cityId` and `councilMeetingId` filter parameters and applies them to `stillPollingMeetings`, `recentPolls`, and the new `conflicts` query. However, the `discoveries` query has no filter at all:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: { taskId: { not: null } },
       // ← no cityId / councilMeetingId filter
       ...
   });
   ```

   This means the **Discoveries** table in the admin UI always renders the full unfiltered history regardless of which city or meeting is selected in the filter dropdowns. Given that the PR explicitly "improves city/meeting filter dropdown sources", this inconsistency is likely unintentional.

   Consider applying the same filters used elsewhere:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: {
           taskId: { not: null },
           ...(cityId && {
               subject: {
                   cityId,
                   ...(councilMeetingId && { councilMeetingId }),
               },
           }),
       },
       ...
   });
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/lib/tasks/pollDecisions.ts
   Line: 197-222

   Comment:
   **`discoveries` ignores city/meeting filter parameters**

   `getPollingStats` accepts `cityId` and `councilMeetingId` filter parameters and applies them to `stillPollingMeetings`, `recentPolls`, and the new `conflicts` query. However, the `discoveries` query has no filter at all:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: { taskId: { not: null } },
       // ← no cityId / councilMeetingId filter
       ...
   });
   ```

   This means the **Discoveries** table in the admin UI always renders the full unfiltered history regardless of which city or meeting is selected in the filter dropdowns. Given that the PR explicitly "improves city/meeting filter dropdown sources", this inconsistency is likely unintentional.

   Consider applying the same filters used elsewhere:

   ```typescript
   const discoveries = await prisma.decision.findMany({
       where: {
           taskId: { not: null },
           ...(cityId && {
               subject: {
                   cityId,
                   ...(councilMeetingId && { councilMeetingId }),
               },
           }),
       },
       ...
   });
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<sub>Last reviewed commit: 49ba069</sub>

### CODE REVIEW

File: src/lib/tasks/pollDecisions.ts [hunk 1]
```diff
@@ -715,8 +893,13 @@ export async function handlePollDecisionsResult(taskId: string, result: PollDeci
                 },
             });
             processedCount++;
+
+            // Track this ADA so later matches in the same batch see it as taken
+            if (match.ada) {
+                adaToExistingSubject.set(match.ada, match.subjectId);
+            }
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil/compare/efafd8ff1b85adf69afe3cf72f49bdceb29a02c6..b6381324c6ba6962c5a14df4ac988402da05bfd7) — `claimedAda` is now cleared after each successful `decision.upsert` via a conditional `updateMany`

File: src/lib/tasks/pollDecisions.ts [hunk 2]
```diff
@@ -687,14 +850,29 @@ export async function handlePollDecisionsResult(taskId: string, result: PollDeci
...
 
+            // Check for ADA conflict: ADA exists on a different subject and isn't being reassigned
+            if (match.ada) {
+                const existingSubjectId = adaToExistingSubject.get(match.ada);
+                if (existingSubjectId && existingSubjectId !== match.subjectId && !reassignedAdas.has(match.ada)) {
+                    // Record the claim on the incoming subject, skip the upsert
+                    await tx.subject.update({
+                        where: { id: match.subjectId },
+                        data: { claimedAda: match.ada },
+                    });
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: False positive — overwriting with the latest claim is intentional. The most recent poll result is the most accurate match. An integration test now documents this behavior explicitly.

File: src/lib/tasks/pollDecisions.ts [hunk 3]
```diff
@@ -632,6 +723,82 @@ export async function getLastPollTimeForMeeting(
...
+    const subject = await prisma.subject.findUnique({
+        where: { id: subjectId },
+        select: { id: true, cityId: true, claimedAda: true },
+    });
+
+    if (!subject) {
+        throw new Error("Subject not found");
+    }
+
+    await withUserAuthorizedToEdit({ cityId: subject.cityId });
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: False positive — subject IDs are CUIDs (not enumerable) and already public in page URLs. The `cityId` is needed from the DB to check city-level authorization — this is the standard auth pattern across the codebase.

**Comment by kouloumos** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil/compare/66de2db684298b4658d7db6cf2855b1fbdf3978a..5df18d6f3fcc2a1d2099bef2538e3573106983f2) to address bot review comment ([1](https://github.com/schemalabz/opencouncil/pull/253#discussion_r2895157932)):

- Catch P2002 on `decision.upsert` during concurrent polls, falling back to `claimedAda` conflict instead of rolling back the entire transaction

File: src/lib/tasks/pollDecisions.ts [hunk 4]
```diff
@@ -687,36 +872,78 @@ export async function handlePollDecisionsResult(taskId: string, result: PollDeci
...
+                    await tx.subject.update({
+                        where: { id: match.subjectId },
+                        data: { claimedAda: match.ada },
+                    });
+                    conflictCount++;
+                    console.log(`ADA conflict (concurrent): ${match.ada} claimed by subject ${match.subjectId}`);
+                    continue;
+                }
+                throw e;
+            }
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/tasks/pollDecisions.ts]: False positive — Prisma interactive transactions use savepoints per operation. When `tx.decision.upsert` fails with P2002, the savepoint is rolled back but the transaction remains usable. The subsequent `tx.subject.update` operates on a fresh savepoint. Even hypothetically without savepoints, the behavior degrades to the pre-fix state (full rollback) — so the code is strictly no worse.
